### PR TITLE
feat(publish): prefill prior metadata and update published workflows in place

### PR DESCRIPTION
**Pull request by MaanilVerma**
> **Goal: make updating a published Hub workflow painless in admin panel.** Today a admin who needs to change a
> published workflow has to reject + reshare + republish, which breaks the share link, resets stats,
> and blanks the thumbnail — and the rejected backlog can't be cleared. This is one of 3 PRs that fix it:
>
> | PR | Repo | Fixes |
> |----|------|-------|
> | [#4505](https://github.com/Comfy-Org/cloud/pull/4505) | `cloud` | Re-publish stops losing data; the editor can read its own published metadata; reviewer search + admin delete |
> | #13139 | `ComfyUI_frontend` | The publish dialog prefills prior metadata + thumbnail and updates in place |
> | [#10 ](https://github.com/Comfy-Org/comfy-admin-panels/pull/10)| `comfy-admin-panels` | Admins can delete rejected/stale workflows |
>
> **Merge order:** `cloud` first (the others depend on its endpoints), then `ComfyUI_frontend` and
> `comfy-admin-panels` in either order.

## Summary

Editor side of the same effort. Re-opening the publish dialog for an already-published workflow now
prefills its description, tags, and thumbnail, restores the thumbnail across step/type changes, keeps
the local workflow name in sync, and labels the action "Update" instead of "Publish". Depends on the
`cloud` PR — that's what finally returns the `share_id` the dialog reads from.

## Changes

- **What**:
  - **Prefill the thumbnail** — `extractPrefill` dropped `thumbnail_url` / `thumbnail_comparison_url`,
    so only the thumbnail *type* was remembered, never the image. Thread the URLs into `PublishPrefill`
    and restore them; on submit, send the existing URL when no new file is attached (reuses the
    `sampleImageUrls` precedent — an existing URL, not a `File`), so re-publishing doesn't blank it.
  - **Uploads survive navigation** — the thumbnail step kept its `File` in local component state, so
    leaving the step and coming back blanked a fresh upload. The step is now controlled — files live in
    the form data, the single source of truth that survives the remount.
  - **Type-gate the prefilled image** — a restored image must only show on the tab it belongs to;
    `existingThumbnailType` keeps an image off the video tab (which was hiding the upload prompt) while
    still restoring it when you toggle back.
  - **Refetch on rename** — the dialog is a reused singleton, so `onMounted` fires once; a watch on the
    active workflow path refetches prefill when a rename changes it (the description was going stale).
  - **Name sync** — editing the name field published a new Hub display name but never renamed the local
    workflow, so the editor tab (and a reload) kept the old name. Publish now renames the local file
    when the chosen name differs.
  - **"Update" CTA** — the intro panel and footer read "Update" (not "Publish") when the workflow is
    already published, and note that the share link + stats are preserved.

## Review Focus

- `existingThumbnailType` is the load-bearing bit for both the preview and submit gating — confirm an
  image prefill never submits as a video after a type toggle, and that toggling back restores it.
- Name sync renames *after* a successful publish and is non-fatal on failure (toast + keep the publish).
  The Hub record is keyed by workflow ID, so the rename doesn't orphan it — worth a sanity check.

## Screenshots

https://github.com/user-attachments/assets/99dd9eff-987f-4ddb-9cf1-e9b40f61e7dc

### CODE REVIEW

File: src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue [hunk 1]
```diff
@@ -205,13 +207,35 @@ function handleRequireProfile() {
...
+}
+
 async function handlePublish(): Promise<void> {
   if (isPublishing.value) {
     return
   }
 
   isPublishing.value = true
   try {
+    await syncWorkflowName()
```
Comment (dante01yoon) [inline src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue]: issue: Renaming before the publish request risks orphaning the existing Hub record. `submitToComfyHub()` reads `workflowStore.activeWorkflow?.path` for `workflowFilename`, and `getPublishStatus()` is also path-keyed. Because `syncWorkflowName()` runs here (238) before `submitToComfyHub()` (239), the publish goes out under the *renamed* path. If the backend matches the existing record by filename/path, the intended update-in-place becomes a fresh record and loses the share link + stats — the exact thing this feature is meant to preserve.

It also opens a partial-failure window: if the rename succeeds but `submitToComfyHub()` then throws, the local file is already renamed while the Hub is untouched, the user sees only a generic "Publish failed" toast, and the next status refresh (now keyed to the new path) can report the workflow as unpublished. Could we submit first, then rename on success (matching the PR description)? The two new ordering tests would flip accordingly.

File: src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue [hunk 2]
```diff
@@ -267,6 +296,14 @@ onMounted(() => {
   void fetchPublishPrefill()
 })
 
+watch(
+  () => workflowStore.activeWorkflow?.path,
+  (newPath, oldPath) => {
+    if (!newPath || newPath === oldPath) return
+    void fetchPublishPrefill()
```
Comment (dante01yoon) [inline src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue:306]: issue: This path watch fires as a side effect of the rename during publish. When `handlePublish()` calls `syncWorkflowName()` (line 238) and the active workflow path changes, this watcher runs `fetchPublishPrefill()` on the next flush — concurrently with the in-flight `submitToComfyHub()`. That issues a redundant `getPublishStatus(newPath)`, can run `applyPrefill()` mid-submit, and flips `isAlreadyPublished` based on a path the backend may not know yet (flickering the footer/intro "Update" label back to "Publish"). The submitted `formData.value` snapshot itself isn't corrupted (applyPrefill replaces the ref with a new object), so it's not data loss — but it's a wasteful, racey refetch. Reordering per the comment above, or guarding this watch with `if (isPublishing.value) return`, would avoid it.

File: src/platform/workflow/sharing/schemas/shareSchemas.ts [hunk 3]
```diff
@@ -10,9 +10,14 @@ export const zPublishRecordResponse = z.object({
   assets: z.array(zAssetInfo).optional()
 })
 
+const zPrefillTag = z
+  .object({ name: z.string(), display_name: z.string() })
```
Comment (dante01yoon) [inline src/platform/workflow/sharing/schemas/shareSchemas.ts:14]: issue: A single malformed tag silently discards the *entire* prefill. `z.array(zPrefillTag)` rejects the whole array if any element is neither `{ name, display_name }` (both required) nor a bare string — e.g. a tag object missing `display_name`. Because `decodeHubWorkflowPrefill()` does `zHubWorkflowPrefillResponse.safeParse(...)` and returns `null` on any failure, one bad tag also drops `description`, `thumbnail_url`, `thumbnail_comparison_url`, and `sample_image_urls`, so re-opening the publish dialog shows nothing prefilled. Per the repo guidance (schema is the spec; sanitize/failover at the consumer), consider tolerating a bad tag — e.g. `.catch(undefined)` per element then filter, or parse tags leniently — so a non-critical tag can't nuke the whole prefill.

File: src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.vue [hunk 4]
```diff
@@ -256,53 +271,83 @@ const thumbnailOptions = [
...
-const thumbnailPreviewUrl = useObjectUrl(thumbnailFile)
-const isVideoFile = ref(false)
+const thumbnailFileUrl = useObjectUrl(() => thumbnailFile ?? undefined)
+const thumbnailPreviewUrl = computed(
+  () => thumbnailFileUrl.value ?? existingSingleUrl.value
+)
+function isVideoUrl(url: string): boolean {
+  return /\.(mp4|webm|mov|m4v|ogv)(\?|#|$)/i.test(url)
+}
+const showVideoPreview = computed(() => {
```
Comment (dante01yoon) [inline src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.vue:297]: question: `showVideoPreview` infers the media kind from the URL extension via `isVideoUrl()`. For restored thumbnails the MIME signal is gone, so a video-mode URL without a recognized extension (presigned/query-only URLs, or an extensionless object key) returns `false` and would render video content inside `` (broken preview). (`.gif`/`.webp` with real extensions are handled correctly here — `isVideoUrl` returns false and your test covers it — so this is not the GIF case.) Are backend `thumbnail_url`s guaranteed to carry a known media extension? If not, persisting the media kind alongside the URL would be more robust than sniffing the path.

File: src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.vue [hunk 5]
```diff
@@ -256,53 +271,83 @@ const thumbnailOptions = [
...
   return !!thumbnailPreviewUrl.value
 })
 
 function clearAllPreviews() {
   if (thumbnailType === 'imageComparison') {
-    comparisonBeforeFile.value = null
-    comparisonAfterFile.value = null
     emit('update:comparisonBeforeFile', null)
     emit('update:comparisonAfterFile', null)
+    emit('update:thumbnailUrl', null)
```
Comment (dante01yoon) [inline src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.vue:345]: question: Clearing a prefilled thumbnail may not actually remove it on update. `clearAllPreviews()` nulls the files/URLs but leaves `existingThumbnailType` set, so on submit `keepsExistingThumbnail` stays true while `thumbnailUrl`/`comparisonAfterUrl` are now null — making `thumbnailTokenOrUrl`/`thumbnailComparisonTokenOrUrl` resolve to `undefined`. If the backend treats `undefined` as "leave unchanged," a user who clears the thumbnail and clicks Update keeps the old image. Is that intended, or should clearing also reset `existingThumbnailType` (and/or send an explicit removal signal)?

**Review by dante01yoon — CHANGES_REQUESTED** [review]
Strict/adversarial pass on the editor-side re-publish change. The mechanics are well-factored and the thumbnail-restore tests are thorough, but I found one data-integrity issue around the rename↔publish ordering (CodeRabbit flagged the same root cause), plus a schema-brittleness issue and a couple of update-mode edge cases worth confirming. Not approving per request.

Things I verified are fine: the controlled-thumbnail refactor correctly threads files+URLs through `formData` so uploads survive remount; the `keepsExistingThumbnail` gating in `useComfyHubPublishSubmission` resolves new-file-vs-existing-URL correctly for image/video/comparison including mixed before/after; `applyPrefill` guards against clobbering user edits; `normalizeTags` round-trips the new `display_name` tags idempotently.

Unanchored notes:
- The PR description says "Name sync renames *after* a successful publish," but the code (and the two new ordering tests) pin the rename *before* `submitToComfyHub`. Please reconcile the description with the intended order — see the inline comment on line 238.
- `codecov/patch` is red at 76% (24 uncovered lines). The new `syncWorkflowName` rename-failure/toast branch and the `fetchPublishPrefill` catch branch appear uncovered; given the rename path is the riskiest new behavior, a failure-path test would be worthwhile.
- nit: in `ComfyHubPublishDialog.test.ts`, `formData: ref((mockFormDataHolder.value = {...}))` does an assignment inside the `ref()` argument as a side effect. A separate `const` for the holder object would read more clearly.

File: src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue (hunk 1 above)
Comment (MaanilVerma) [reply src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue]: Yes, I agree. That’s actually how I had it implemented initially, but I trusted CodeRabbit's review blindly. I'll get this corrected right away.

File: src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.vue (hunk 5 above)
Comment (MaanilVerma) [reply src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.vue:345]: I looked into this, and clearing a thumbnail before clicking "Update" keeps the old image by design. The backend retains the previous image if no new thumbnail is sent, which acts as a helpful safeguard against accidentally wiping it out.

Generally, users clear a thumbnail only to replace it with a new upload, which works perfectly fine as the new image takes over.

In fact, accidental image deletion was an issue brought up previously, where creators would accidentally wipe their thumbnails and admin have to ask them to re-upload them. The current logic ensures we always keep at least the older image so that workflows on the Hub never end up missing a thumbnail.

File: src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.test.ts [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    )
+
+    await user.click(screen.getByTestId('type-image'))
+
+    expect(onUpdateThumbnailType).toHaveBeenCalledWith('image')
+    expect(onUpdateComparisonBeforeFile).toHaveBeenCalledWith(null)
+    expect(onUpdateComparisonAfterFile).toHaveBeenCalledWith(null)
+    expect(onUpdateThumbnailUrl).not.toHaveBeenCalled()
+    expect(onUpdateComparisonAfterUrl).not.toHaveBeenCalled()
+  })
```
[the remark this answers is not in the document]
Comment (MaanilVerma) [reply src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.test.ts:128]: Verified against the code. On a type switch the component intentionally emits update:thumbnailFile: null (files reset; this has been the behavior since the feature shipped, and it's deliberate because the component is controlled — state lives in the parent's formData). So the suggested assertion expect(onUpdateThumbnailFile).not.toHaveBeenCalledWith(null) fails against the real code — it'd be testing behavior that doesn't exist. What is preserved across a type switch is the restored URLs, which the existing test already covers.

UX Angle: 

The reason is that the thumbnail type isn't a setting on one object — it's a choice of what kind of thumbnail you're making: a single image, a video, or a before/after comparison are three different deliverables. When you switch from image to comparison, "your image file" stops being meaningful (comparison has a before and an after, not a single image), so carrying it over would be ambiguous and risk submitting the wrong file for the wrong mode. Starting the new mode's inputs clean keeps the submit logic unambiguous.

What is preserved is restored URLs — because those came from a previously-published state and have a stable meaning, so the component keeps them inert and restores them when you switch back (covered by the test at 96-131). The split is intentional: committed URLs persist, in-progress mode-local files don't.

File: src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.vue (hunk 4 above)
Comment (MaanilVerma) [reply src/platform/workflow/sharing/components/publish/ComfyHubThumbnailStep.vue:297]: Yes it's safe with the current backend , thumbnail URLs always carry a known extension, and the only video formats end in .mp4/.webm, which the check matches. So a restored video thumbnail won't render inside  today.

You're right that this relies on that guarantee holding, though , if video ever came back without a clear extension (presigned or extensionless URLs), the path-sniffing would break, and persisting the media kind alongside the URL would be sturdier. It's pure hardening against a future change of backend if something breaks, rather than a live bug, but still I'm fixing it now.

File: src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue [hunk 7]
```diff
@@ -213,6 +236,7 @@ async function handlePublish(): Promise<void> {
   isPublishing.value = true
   try {
     await submitToComfyHub(formData.value)
+    await syncWorkflowName()
```
Comment (dante01yoon) [inline src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue:239]: question: The reorder correctly fixes the *in-flight* publish, but the rename here still moves the userdata file. Publish status is path-keyed end to end: the FE calls `getPublishStatus(activeWorkflow.path)` → `GET /userdata/{path}/publish`, and on the cloud side `GetPublishStatus` resolves the record via `getWorkflowByName(...name)` → `GetPublishedWorkflowByWorkflowIDAndType(wf.ID, ...)`. So whether an update that *also* renames stays detected as published on the next open depends on whether the cloud userdata rename preserves the workflow row's identity (`wf.ID`) — and neither this PR nor cloud #4505 touches the rename/move handler.

If the rename updates the existing workflow row's name in place (same `wf.ID`), everything's fine. But if a rename creates a new workflow row (or leaves the published row pointing at the old name), then on the next open `getPublishStatus(newPath)` returns 404 → `isAlreadyPublished = false` → the footer shows "Publish", and re-publishing would create a second record under the new path and orphan the original `share_id` + stats — the exact data loss this effort prevents. Prefill is still restored from the local cache, so it's silent.

This only bites the update-then-rename path (publishing an already-published workflow while changing its name). Could you confirm the cloud userdata rename keeps `wf.ID` (or otherwise carries the published record across the move)? If it does, I'm happy to approve.

**Review by dante01yoon — COMMENTED** [review]
Re-review since my CHANGES_REQUESTED at `f790ebf7`. All five of my earlier findings are addressed in code with regression coverage, and CI is fully green. I have one remaining backend-behavior question (inline) that keeps me from a confident approve, so leaving this as a COMMENT.

Resolved (verified against `ab62cd71`):
- **Rename ordering** (`46755269a0`) — `submitToComfyHub()` now runs before `syncWorkflowName()`, so the update goes out under the path the existing record is keyed by; the ordering test now asserts `invocationCallOrder` (submit < rename) and there's a new "rename fails → success toast + warn" test covering the partial-failure window.
- **Path-watch race** (`46755269a0` + `36069a7f`) — the path watcher early-returns on `isPublishing`, and `fetchPublishPrefill` drops stale responses after each await (`activeWorkflow.path !== path`). No more mid-submit refetch / Update→Publish flicker.
- **Malformed prefill tag** (`1d0346a2`) — `zPrefillTagList` now does `zPrefillTag.optional().catch(undefined)` per element then filters, so one bad tag no longer nukes the whole prefill (description/thumbnail/samples survive).
- **Extensionless video preview** (`ab62cd71`) — `showVideoPreview` now uses `!isAnimatedImageUrl(url)`, so a video-mode URL without a recognized image extension defaults to `<video>` while real `.gif`/`.webp` still render as ``. This also covers CodeRabbit's residual video-mode GIF/WebP flag.
- **Clearing a prefilled thumbnail** — confirmed intended by design (backend retains the prior image as an accidental-wipe safeguard); the added-then-dropped reset commits net to no behavioral change. Thanks for the detailed rationale.

The one open question is inline on the `syncWorkflowName()` call.

File: src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue (hunk 7 above)
Comment (MaanilVerma) [reply src/platform/workflow/sharing/components/publish/ComfyHubPublishDialog.vue:239]: Yes, @dante01yoon  , the rename keeps wf.ID, so the published record survives.

A rename within workflows/ calls RenameWorkflow, which does UpdateOneID(wf.ID).SetName(newName) , in-place, same row ID. The published_workflow record is keyed by workflow_id, not name, so it stays linked: GetPublishStatus(newPath) finds the renamed row and looks up the record by its unchanged ID. So it stays detected as published and re-publishing updates in place , no orphaned share_id/stats.

**Review by dante01yoon — APPROVED** [review]
Approving on head `a004f591`.

My one remaining open question — whether an update that *also* renames stays detected as published — is resolved. Confirmed that the cloud userdata rename is an in-place `UpdateOneID(wf.ID).SetName(newName)` (same row ID) and `published_workflow` is keyed by `workflow_id`, not name, so `GetPublishStatus(newPath)` finds the renamed row and looks up the record by its unchanged ID. The share link + stats carry across the move, which was the exact condition I needed.

All five of my earlier findings remain resolved with regression coverage (submit-before-rename ordering, `isPublishing` path-watch guard, stale-prefill drop, per-element tag tolerance, extensionless video-mode preview). No behavioral source changes since my last pass — the two new commits are merge-from-main only. CI is fully green.

Thanks for the thorough back-and-forth on the rename↔publish ordering and the cloud-side confirmation.
